### PR TITLE
Notifications

### LINKED ISSUE #85: Notifications
1. Send an email on activation of job using sequelize hooks.
We can write a combination of sql and ts in hooks.
First find the mails of the students eligible and then mail them.

2. Send an email on change of application event id to student.
Again use sequelize hooks to acheive it.

3. Send an email to student on offer table entry.

4. Send an email on filling of jaf to admin using sequelize hooks to detect the creation of job ( only by jaf filling ).

### PR DESCRIPTION
**Pull request by AryanGKulkarni**
Closes #85
Closes #91 

Mails will be sent on:

- [x] Offer Table Entry
- [x] Creation of job through JAF filling
- [x] After updating job status to true
- [x] Change in eventId of ApplicationModel
- [x] Token email on login (note: I have just changed the name of function sendEmail to sendTokenEmail in auth.controller.ts)
  

Use following environment variables:

MAIL_HOST=smtp.gmail.com
MAIL_PORT=587
MAIL_USER= email id of the sender
MAIL_PASSWORD= app password generated by gmail from https://security.google.com/settings/security/apppasswords
DEFAULT_MAIL_FROM= email id of the sender
APP_NAME=TPC Portal

### CODE REVIEW

File: src/db/models/ApplicationModel.ts [hunk 1]
```diff
@@ -75,4 +90,62 @@ export class ApplicationModel extends Model<ApplicationModel> {
...
+        // Retrieve the associated student details
+        const student = await StudentModel.findByPk(application.studentId);
+        if (!student) {
+          console.error(`Student not found for offer ${application.id}`);
+        }
+
+        // Find the user associated with this student
+        const user = await UserModel.findByPk(student.userId);
+        if (!user) {
+          console.error(`User not found for student ${student.id}`);
```
Comment (gjain-7) [inline src/db/models/ApplicationModel.ts:134]:
```suggestion
        const student = await StudentModel.findByPk(application.studentId, {
          include: [{
          model: UserModel,
          as: "user",
        }]
        });
        if (!student) {
          throw new NotFoundException(`Student not found for offer ${application.id}`);
        }
        
        if (!student.user) {
          throw new NotFoundException(`User not found for student ${student.id}`);
```

File: src/db/models/ApplicationModel.ts [hunk 2]
```diff
@@ -75,4 +90,62 @@ export class ApplicationModel extends Model<ApplicationModel> {
...
+        }
+
+        // Prepare the email data
+        const data: SendEmailDto = {
+          from: { name: "TPC Portal", address: "aryangkulkarni@gmail.com" },
```
Comment (gjain-7) [inline src/db/models/ApplicationModel.ts:139]: read for .env here

File: src/services/EmailService.ts [hunk 3]
```diff
@@ -1,31 +1,88 @@
...
+import { Address } from "nodemailer/lib/mailer";
+
+export type SendEmailDto = {
+  from?: Address;
+  recepients: Address[];
+  subject: string;
+  html: string;
+  text?: string;
+  placeholderReplacements?: Record<string, string>;
+};
```
Comment (gjain-7) [inline src/services/EmailService.ts:19]:
```suggestion
export class SendEmailDto = {
  from?: Address;
  recepients: Address[];
  subject: string;
  html: string;
  text?: string;
  placeholderReplacements?: Record<string, string>;
};
```

File: src/services/EmailService.ts [hunk 4]
```diff
@@ -1,31 +1,88 @@
+/* eslint-disable no-console */
 // email.service.ts
 
 import { Global, HttpException, Injectable, Logger } from "@nestjs/common";
 import { globalAgent } from "http";
-import * as nodemailer from "nodemailer";
+const nodemailer = require("nodemailer");
```
Comment (gjain-7) [inline src/services/EmailService.ts:6]: why `require`?

File: src/services/EmailService.ts [hunk 5]
```diff
@@ -1,31 +1,88 @@
+/* eslint-disable no-console */
```
Comment (gjain-7) [inline src/services/EmailService.ts:1]:
```suggestion
```

File: src/db/models/JobModel.ts [hunk 6]
```diff
@@ -177,4 +198,78 @@ export class JobModel extends Model<JobModel> {
...
+  static async sendEmailHook(instance: JobModel) {
+    console.log("New entry created");
+    const mailerService = new EmailService();
+
+    const dto: SendEmailDto = {
+      from: { name: "TPC Portal", address: "aryangkulkarni@gmail.com" },
+      recepients: [{ address: "me210003016@iiti.ac.in" }], // Put your email address for testing
+      subject: "Test email",
+      html: "<p>New job entry was created</p>",
+    };
```
Comment (gjain-7) [inline src/db/models/JobModel.ts:212]: read from .env

put better content.

File: src/db/models/JobModel.ts [hunk 7]
```diff
@@ -177,4 +198,78 @@ export class JobModel extends Model<JobModel> {
...
+              salary.categories.includes(student.category) &&
+              salary.genders.includes(student.gender) &&
+              student.tenthMarks >= salary.tenthMarks &&
+              student.twelthMarks >= salary.twelthMarks
+            ) {
+              console.log("student added", student.id);
+              eligibleStudents.add(student);
+            }
+          });
+        }
```
Comment (gjain-7) [inline src/db/models/JobModel.ts:255]: use db query instead, more efficient.
you can just use `filterSalaries` which does that.
use case here https://github.com/Web-Team-IITI-Gymkhana/tpc-backend/blob/3a80c7e470c220c2bb35ad61cba93c6e342b3b5b/src/student-view/salary/salary.service.ts#L51

also include this condition
https://github.com/Web-Team-IITI-Gymkhana/tpc-backend/blob/3a80c7e470c220c2bb35ad61cba93c6e342b3b5b/src/student-view/salary/salary.service.ts#L64-L68

File: src/db/models/JobModel.ts [hunk 8]
```diff
@@ -177,4 +198,78 @@ export class JobModel extends Model<JobModel> {
...
+
+        // console.log("eligible Students ", eligibleStudents);
+
+        // Prepare and send the email data
+        for (const student of eligibleStudents) {
+          const user = await UserModel.findByPk(student.userId);
+          const data: SendEmailDto = {
+            from: { name: "TPC Portal", address: "aryangkulkarni@gmail.com" },
```
Comment (gjain-7) [inline src/db/models/JobModel.ts:263]: .env

File: src/db/models/JobModel.ts [hunk 9]
```diff
@@ -177,4 +198,78 @@ export class JobModel extends Model<JobModel> {
...
+              eligibleStudents.add(student);
+            }
+          });
+        }
+
+        // console.log("eligible Students ", eligibleStudents);
+
+        // Prepare and send the email data
+        for (const student of eligibleStudents) {
+          const user = await UserModel.findByPk(student.userId);
```
Comment (gjain-7) [inline src/db/models/JobModel.ts:261]: while querying include the user in student model to save these db queries.

File: src/db/models/OffCampusOfferModel.ts [hunk 10]
```diff
@@ -82,4 +86,39 @@ export class OffCampusOfferModel extends Model<OffCampusOfferModel> {
...
+    const mailerService = new EmailService();
+
+    // Iterate over each newly created instance
+    for (const offer of instance) {
+      // Find the student associated with this offer
+      const student = await StudentModel.findByPk(offer.studentId);
+      if (!student) {
+        console.error(`Student not found for offer ${offer.id}`);
+        continue; // Skip to the next offer if student not found
+      }
```
Comment (gjain-7) [inline src/db/models/OffCampusOfferModel.ts:102]:
```suggestion
    const students = await StudentModel.findAll({
      where: {
        id: instance.map(offer => offer.studentId)
        // + include whatever needed (like user)
      }
    })
```

File: src/db/models/OffCampusOfferModel.ts [hunk 11]
```diff
@@ -82,4 +86,39 @@ export class OffCampusOfferModel extends Model<OffCampusOfferModel> {
...
+      const user = await UserModel.findByPk(student.userId);
+      if (!user) {
+        console.error(`User not found for student ${student.id}`);
+        continue; // Skip to the next offer if user not found
+      }
+
+      // Prepare the email data
+      const dto: SendEmailDto = {
+        from: { name: "TPC Portal", address: "aryangkulkarni@gmail.com" },
+        // recepients: [{ address: "me210003016@iiti.ac.in" }],  // Put your email address for testing
```
Comment (gjain-7) [inline src/db/models/OffCampusOfferModel.ts:114]: same

File: src/db/models/OnCampusOfferModel.ts [hunk 12]
```diff
@@ -55,4 +59,39 @@ export class OnCampusOfferModel extends Model<OnCampusOfferModel> {
...
+      // Find the user associated with this student
+      const user = await UserModel.findByPk(student.userId);
+      if (!user) {
+        console.error(`User not found for student ${student.id}`);
+        continue; // Skip to the next offer if user not found
+      }
+
+      // Prepare the email data
+      const data: SendEmailDto = {
+        from: { name: "TPC Portal", address: "aryangkulkarni@gmail.com" },
```
Comment (gjain-7) [inline src/db/models/OnCampusOfferModel.ts:86]: same

**Review by gjain-7 — CHANGES_REQUESTED** [review]
Thanks! Suggestions inline.

1. Remove hardcoded strings and import from config
2. Remove console.log statements
3. Remove unnecessary comment and eslint comments

File: src/db/models/ApplicationModel.ts [hunk 13]
```diff
@@ -75,4 +100,59 @@ export class ApplicationModel extends Model<ApplicationModel> {
...
+        const mailerService = new EmailService();
+
+        const student = await StudentModel.findByPk(application.studentId, {
+          include: [
+            {
+              model: UserModel,
+              as: "user",
+            },
+          ],
+        });
```
Comment (cse210001015) [inline src/db/models/ApplicationModel.ts:138]:
```suggestion
        const students = await StudentModel.findAll({studentId: applications.map((application) => application.studentId)}, {
          include: [
            {
              model: UserModel,
              as: "user",
            },
          ],
        });
```

**Review by gjain-7 — APPROVED** [review]

**Review by gjain-7 — APPROVED** [review]
Just remove the N+1 queries (db query in for loop) in favour of batch queries.

**Pull request merged**